### PR TITLE
MAINT: `stats.mstats` consistency with `stats.stats`

### PR DESCRIPTION
**Pull request by aeklant**
make the signatures of common functions between `stats`
and `mstats` as similar as possible.

### CODE REVIEW

**Comment by rgommers** [comment]
Extending the test for ``tmax`` so this change is covered would be useful. Can be completely analogous to the ``tmin`` test.

**Comment by rgommers** [comment]
I assume this is a WIP PR and you want to add a few more functions? Would be good to signal when it's ready to merge, and change the PR title to be a bit more specific.

**Comment by aeklant** [comment]
@rgommers, you beat me to it. Yes, this is a WIP, I just added the most trivial function to have this PR as a placeholder. 

I am writing up the list of functions along with my observations. All in all I found 13 functions that need to be addressed. Will post them presently for discussion.

**Comment by aeklant** [comment]
Here is the list of common functions that have some differences coupled with some of my observations:

`ttest_ind`:
This one is addessed in gh-5007. So nothing to do here.

`describe`:
Also (partially) addressed in gh-5007. `ddof` default is different so it must be decided if it should be changed to make them identical. 

`tmax`:
`upperlimit` in `mstats` had no default. Fixed. Test pending.

`tmean`, `tvar` and `tsem`:
These functions (in `mstats`) use a helper function called `trima` which doesn't implement `axis`, so none of these functions do, either. `tvar` and `tsem` are also missing the `ddof` keyword.

`mannwhitneyu`:
Missing `alternative` keyword. This function is scheduled for deprecation so I'm not sure whether to change it or not.

`kendalltau`: `initial_lexsort` keyword is missing in `mstats`. `use_ties` and `use_missing` keywords are missing in `stats` 

`trimboth`: `proportiontocut` has no default value in `stats`. `axis` keyword has a different default value (0 and 1). `inclusive` keyword missing from `stats`. Same input with different names (`a` and `data`).

`linregress`: In `mstats` the input params are `*args` but the implementation assumes just one or two params (equivalent to the `stats` version: `x, y=None`) and then it defers to calculation to the `stats.linregress` so the function signature should be changed accordingly.

`scoreatpercentile`: `interpolation_method` keyword missing from `mstats`. `alphap` and `betap` keywords missing from `stats`. Same input parameter with different names (i.e. `a` and `data`).

`signaltonoise`: `ddof` keyword missing from `mstats`. Same input parameter with different names (i.e. `a` and `data`).

`spearmanr`: `axis` keyword missing from `mstats`. `use_ties` keyword missing from `stats`. Same input parameter with different names (i.e. `a, b` and `x, y`).

**Comment by aeklant** [comment]
General considerations:

* What to do when an input is implemented in both `stats` and `mstats` but with different names? Should the difference be allowed or should we risk breaking some code for the sake of similarity? **note: these are index input values, not keyword input values.
* What about when one input has different default values? These changes are bound to break some code.
* Should `mannwhitneyu` be ignored due to being scheduled for deprecation? 
* Missing keywords should be reviewed on a case to case basis IMO so I will not comment on those yet.

**Comment by aeklant** [comment]
Notes:
* Added a test for `tmax`.
* Added a default value of `0.2` to keyword `proportiontocut` in `stats.trimboth`.
* Changed `(*args)` for `(x, y=None)` in `mstats.linregress` and removed some code that didn't seem to make sense (e.g. `if args.size == 2` nested inside `if args.size == 1`).

I'm afraid for all of the other changes that I outlined, there will need to be a discussion on a case to case basis. I wonder if it would be best to treat those in separate PRs.

**Comment by ev-br** [comment]
Fot completeness, here's https://github.com/scipy/scipy/issues/3328 whoch is also dealing with stats/mstats inconsistency

**Comment by rgommers** [comment]
I'd say add to this PR only the changes that require no discussion or backwards compat issues. I think the only ones in addition to the changes you already have are:

- ``tmean``, ``tvar`` and ``tsem``: can add ``axis`` and ``ddof`` keywords to all of them.

**Comment by rgommers** [comment]
For ``signaltonoise`` and ``mannwhitneyu``, those are deprecated and about to be deprecated respectively, so don't bother with those.

**Comment by rgommers** [comment]
So then there are exactly 5 (if I counted correctly) functions left where we'd have to make a backwards incompatible change:

- describe
- kendalltau
- trimboth
- scoreatpercentile
- spearmanr

I'd say open a new issue for those; it'll get lost in this PR. I'd be tempted to just break backwards compat for all of them in the ``mstats`` versions.

File: scipy/stats/mstats_basic.py [hunk 1]
```diff
@@ -642,19 +642,9 @@ def linregress(*args):
...
-            x = args[0]
-            y = args[1]
-        else:
-            x = args[:, 0]
-            y = args[:, 1]
-    else:
-        # Input is two 1-D arrays
-        x = ma.array(args[0]).flatten()
-        y = ma.array(args[1]).flatten()
+    x = ma.array(x).flatten()
```
Comment (rgommers) [inline scipy/stats/mstats_basic.py]: This doesn't look right for a single 2-D input array. It should be split in two, like the old code did. Can you add a test for this and undo the change here?

File: scipy/stats/stats.py [hunk 2]
```diff
@@ -2492,7 +2492,7 @@ def sigmaclip(a, low=4., high=4.):
     return SigmaclipResult(c, critlower, critupper)
 
 
-def trimboth(a, proportiontocut, axis=0):
+def trimboth(a, proportiontocut=0.2, axis=0):
```
Comment (rgommers) [inline scipy/stats/stats.py]: This one is tricky, because it doesn't really make sense to have a default here. I'd rather drop the default from the mstats version, even though that's backwards-incompatible.

Comment (rgommers) [inline scipy/stats/stats.py]: We're anyway making other backwards incompatible changes there (at least I'd like to), so let's just add this one to that. Shouldn't be a big deal.

**Comment by aeklant** [comment]
The `stats` version of `tmean`, `tvar` and `tsem` already supports masked arrays so reimplementing seems redundant.

Also, the `mstats` version all do their work through a helper function `trima`. It would be easier to have `mstats` functions delegate to `stats`I think.

**Comment by rgommers** [comment]
> The stats version of tmean, tvar and tsem already supports masked arrays so reimplementing seems redundant.

I don't think so. The first line is ``a = asarray(a)``, so the mask simply gets thrown away in the ``stats`` version. To check that that's indeed the case:

    In [10]: x = np.ma.arange(3)

    In [11]: stats.tmean(x)
    Out[11]: 1.0

    In [12]: x.mask = [0, 0, 1]

    In [13]: stats.tmean(x)
    Out[13]: 1.0

    In [14]: stats.mstats.tmean(x)
    Out[14]: 0.5

**Comment by aeklant** [comment]
Latest changes:

* Addition of `axis` keyword to `tmean`, `tsem` and `tvar`.
* Addition of `ddof` keyword to `tsem`and `tvar`.
* Removed default value for `proportiontocut` in `stats.trimboth`.
* Eliminated the change flatten x in `linregress`.

Notes:
* The implementation of `tvar` was changed because the previous implementation lost precision when adding the `axis` and `ddof` keywords.
* The implementation of flatten `x` and `y` in `linregress` had to be changed because of the modification of its signature (see gh-5168)

If there are any comments or changes I will try to address them as soon as possible, but I am officially signing off. I am moving to Canada in less than 24 hours and I have some serious packing to do and documents to organise.

**Comment by rgommers** [comment]
@aeklant have a good flight!

**Comment by rgommers** [comment]
okay, this all looks good. merging. thanks @aeklant
